tapdb/universe_federation.go

```diff
@@ -267,6 +299,261 @@ func (u *UniverseFederationDB) LogNewSyncs(ctx context.Context,
+	leafKeyOutpointBytes, err := encodeOutpoint(leafKey.OutPoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// Encode the leaf script key pub key as bytes. We'll use this to look
+	// up the leaf ID in the DB.
+	scriptKeyPubKeyBytes := schnorr.SerializePubKey(
+		leafKey.ScriptKey.PubKey,
+	)
+
+	var (
+		readTx        = NewUniverseFederationReadTx()
+		proofSyncLogs []*universe.ProofSyncLogEntry
+	)
+
+	err = u.db.ExecTx(ctx, &readTx, func(db UniverseServerStore) error {
+		params := QueryFedProofSyncLogParams{
+			SyncDirection:         sqlStr(string(syncDirection)),
+			Status:                sqlStr(string(syncStatus)),
+			LeafNamespace:         sqlStr(uniID.String()),
+			LeafMintingPointBytes: leafKeyOutpointBytes,
+			LeafScriptKeyBytes:    scriptKeyPubKeyBytes,
+		}
+		logEntries, err := db.QueryFederationProofSyncLog(ctx, params)
+
+		// Parse database proof sync logs. Multiple log entries may
+		// exist for a given leaf because each log entry is unique to a
+		// server.
+		for idx := range logEntries {
+			entry := logEntries[idx]
+
+			parsedLogEntry, err := parseSqlProofSyncLogEntry(
+				ctx, entry, db,
+			)
+			if err != nil {
+				return err
+			}
+
+			proofSyncLogs = append(proofSyncLogs, parsedLogEntry)
+		}
+
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return proofSyncLogs, nil
+}
+
+// FetchPendingProofsSyncLog queries the federation proof sync log and returns
+// all log entries with sync status pending.
+func (u *UniverseFederationDB) FetchPendingProofsSyncLog(ctx context.Context,
+	syncDirection *universe.SyncDirection) ([]*universe.ProofSyncLogEntry,
+	error) {
+
+	var (
+		readTx        = NewUniverseFederationReadTx()
+		proofSyncLogs []*universe.ProofSyncLogEntry
+	)
+
+	err := u.db.ExecTx(ctx, &readTx, func(db UniverseServerStore) error {
+		// If the sync direction is not set, then we'll query for all
+		// pending proof sync log entries.
+		var sqlSyncDirection sql.NullString
+		if syncDirection != nil {
+			sqlSyncDirection = sqlStr(string(*syncDirection))
+		}
+
+		sqlProofSyncStatus := sqlStr(
+			string(universe.ProofSyncStatusPending),
+		)
+
+		params := QueryFedProofSyncLogParams{
+			SyncDirection: sqlSyncDirection,
+			Status:        sqlProofSyncStatus,
+		}
+		logEntries, err := db.QueryFederationProofSyncLog(ctx, params)
+		if err != nil {
+			return fmt.Errorf("unable to query proof sync log: %w",
+				err)
+		}
+
+		// Parse log entries from database row.
+		for idx := range logEntries {
+			entry := logEntries[idx]
+
+			parsedLogEntry, err := parseSqlProofSyncLogEntry(
+				ctx, entry, db,
+			)
+			if err != nil {
+				return err
+			}
+
+			proofSyncLogs = append(proofSyncLogs, parsedLogEntry)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return proofSyncLogs, nil
+}
+
+// parseSqlProofSyncLogEntry returns a parsed proof sync log entry given a DB
+// row.
+func parseSqlProofSyncLogEntry(ctx context.Context, entry ProofSyncLogEntry,
```

Comment:
Related to the comment above: Since we're doing another DB call here, the function name is a bit misleading, as we're doing way more than just parsing (that's why I missed this in the first review iteration).

Reply:
What name should I use instead? `marshalSqlProofSyncLogEntry`?

I wish we could do `universe.ProofSyncLogEntry::FromDbType(...)`, that way it should be easy to find `FromDbType` on the type namespace.